content/blog/pitfalls-of-safe-rust/index.md

```diff
@@ -0,0 +1,687 @@
+Instead, implement `Debug` manually for types that contain sensitive information.
+
+```rust
+// DON'T: Expose sensitive data in debug output
+struct User {
+    username: String,
+    password: String,  // Will be printed in debug output!
+}
+
+// DO: Protect sensitive data
+struct Password(String);
+
+impl std::fmt::Debug for Password {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("[REDACTED]")
+    }
+}
+
+struct User {
+    username: String,
+    password: Password,
+}
+```
+
+For production code, use a crate like [`secrecy`](https://crates.io/crates/secrecy). 
+
+However, it's not black and white either:
+If you implement `Debug` manually, you might forget to update the implementation when your struct changes.
+A common pattern is to destructure the struct in the `Debug` implementation to catch such errors:
+
+```rust
+impl fmt::Debug for Settings {  
+   fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {  
+       // Destructure the struct to catch changes
+       // This way, the compiler will warn you if you add a new field
+       // and forget to update the Debug implementation
+       let Settings { version, base_url, content_dir } = self;  
+       f.debug_struct("Settings")  
+           .field("version", version)
+           .field("base_url", &base_url.as_str())
+           .field("content_dir", content_dir)
+           .finish()  
+   }  
+}
+```
+
+Thanks to [Wesley Moore (wezm)](https://www.wezm.net) for the hint.
+
+## Implement Secure Serialization
+
+On the same note, be careful with serialization and deserialization.
+The values you read/write might not be what you expect.
+
+```rust
+// DON'T: Blindly derive serialization
+#[derive(Serialize, Deserialize)]
+struct UserCredentials {
+    #[serde(default)]  // Ouch, we could deserialize to empty string!
+    username: String,
+    #[serde(default)]
+    password: String,
+}
+
+// DO: Implement custom serialization with validation
+use serde::{Deserialize, Deserializer};
+
+#[derive(Serialize)]
+struct UserCredentials {
+    username: Username,
+    #[serde(skip_serializing)]  // Never serialize passwords
+    password: Password,
+}
+
+impl<'de> Deserialize<'de> for UserCredentials {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct Raw {
+            username: String,
+            password: String,
+        }
+
+        let raw = Raw::deserialize(deserializer)?;
+        
+        // Validate during deserialization
+        // Throw an error if the username or password is invalid
+        let username = Username::new(&raw.username)
+            .map_err(serde::de::Error::custom)?;
+        let password = Password::new(&raw.password)
+            .map_err(serde::de::Error::custom)?;
+
+        Ok(UserCredentials { username, password })
+    }
+}
+```
```

Comment:
I'm questioning if manually implementing that traits is the best recommendation. Forgoing the derive seems error-prone. 

Furthermore, this issue isn't just with deserialization! I could just create an instance of the struct in code with empty strings and that would be invalid, so it kind of goes back to the earlier point about using types to make invalid things unrepresentable. In that vein, I found this article, which proposes a validation strategy for deserialization. So the idea would be to use a new type like NonemptyString for the fields. https://dev.to/equalma/validate-fields-and-types-in-serde-with-tryfrom-c2n